### DISCUSSION TITLE
How to modify a variable during the backwards pass in Flax?

### DISCUSSION THREAD

**Discussion by marcvanzee** [google/flax#2079] [Q&A] [answered]
Possible motivation: The algorithm relies on keeping track of statistics of activation gradients over multiple training steps.

**Comment by marcvanzee** [comment, marked as answer]
Below is input from @jheek.

Using a lifted `custom_vjp` is an "odd duck". We can't add auxiliary outputs to a grad so we don't have a way to mutate collections during the backward pass. So we freeze the state such that you can access is but not update it

There's a special case that can be solved and that's usually what you want: Tracking intermediate gradients.

So in practise you can do this by splitting up your "stateful backward pass" into 2 parts:

1.  Store normal state in collection during the forward pass 
2. use `x = x + self.variable('grad_trackers', 'name', nn.initializers.zeros)` to any variable `x` for which you need to track the partial derivative.
3. take the grad w.r.t. params & grad_trackers
4. Use grad_trackers and other state to update the backward state (in this case compute backward activation statistics)
